File: abba/stats.js
var Abba = {};

// Friendly wrapper over jStat's normal distribution functions.
Abba.NormalDistribution = function(mean, standardDeviation) {
    if (mean === undefined) {
        mean = 0;
    }
    if (standardDeviation === undefined) {
        standardDeviation = 1;
    }
    this.mean = mean;
    this.standardDeviation = standardDeviation;
};
Abba.NormalDistribution.prototype = {
    density: function(value) {
        return jStat.normal.pdf(value, this.mean, this.standardDeviation);
    },

    // Returns P(x < value) for x standard normal. value may be any number.
    cdf: function(value) {
        return jStat.normal.cdf(value, this.mean, this.standardDeviation);
    },

    // Returns P(x > value) for x standard normal. value may be any number.
    survival: function(value) {
        return 1 - this.cdf(value);
    },

    // Returns z such that P(x < z) = probability for x standard normal.
    // probability must be in (0, 1).
    inverseCdf: function(probability) {
        return jStat.normal.inv(probability, this.mean, this.standardDeviation);
    },

    // Returns z such that P(x > z) = probability for x standard normal.
    // probability must be in (0, 1).
    inverseSurvival: function(probability) {
        return this.mean - (this.inverseCdf(probability) - this.mean);
    },
};

/* Distribution functions for the binomial distribution. Relies entirely on the normal
   approximation.

   jStat's binomial functions do not seem be to reliable or as performant for large cases. This
   class could be improved by making it compute exact binomial functions for small cases and fall
   back to the normal approximation for large cases.
*/
Abba.BinomialDistribution = function(numSamples, probability) {
    this.numSamples = numSamples;
    this.probability = probability;
    this.expectation = numSamples * probability;
    this.standardDeviation = Math.sqrt(this.expectation * (1 - probability));

    // normal approximation to this binomial distribution
    this._normal = new Abba.NormalDistribution(this.expectation, this.standardDeviation);
    this._lowerTailProbability = this._normal.cdf(-0.5);
    this._upperTailProbability = this._normal.survival(numSamples + 0.5);
};
Abba.BinomialDistribution.prototype = {
    mass: function(count) {
        return this._normal.density(count);
    },

    _rescaleProbability: function(probability) {
        return probability / (1 - this._lowerTailProbability - this._upperTailProbability);
    },

    cdf: function(count) {
        if (count < 0) {
            return 0;
        } else if (count >= this.numSamples) {
            return 1;
        } else {
            return this._rescaleProbability(
                this._normal.cdf(count + 0.5) - this._lowerTailProbability);
        }
    },

    survival: function(count) {
        return 1 - this.cdf(count);
    },

    inverseCdf: function(probability) {
        return Math.max(0, Math.min(this.numSamples, this._normal.inverseCdf(probability)));
    },

    inverseSurvival: function(probability) {
        return Math.max(0, Math.min(this.numSamples, this._normal.inverseSurvival(probability)));
    },
};

Abba.ValueWithInterval = function(value, intervalWidth) {
    this.value = value;
    this.intervalWidth = intervalWidth;
}
Abba.ValueWithInterval.prototype = {
    range: function() {
        return {
            lowerBound: this.value - this.intervalWidth,
            upperBound: this.value + this.intervalWidth
        };
    },
};

// A value with standard error, from which a confidence interval can be derived.
Abba.ValueWithError = function(value, error) {
    this.value = value;
    this.error = error;
}
Abba.ValueWithError.prototype = {
    /* criticalZValue should be the value at which the right-tail probability for a standard
       normal distribution equals half the desired alpha = 1 - confidence level:

       P(Z > zValue) = 1 - alpha / 2

       where Z is an N(0, 1) random variable.  Use NormalDistribution.inverseSurvival(), or see
       http://en.wikipedia.org/wiki/Standard_normal_table.
    */
    confidenceIntervalWidth: function(criticalZValue) {
        return criticalZValue * this.error;
    },

    valueWithInterval: function(criticalZValue) {
        return new Abba.ValueWithInterval(this.value, this.confidenceIntervalWidth(criticalZValue));
    },
};

// Represents a binomial proportion with numSuccesses successful samples out of numSamples total.
Abba.Proportion = function(numSuccesses, numSamples) {
    this.numSuccesses = numSuccesses;
    this.numSamples = numSamples;
    this._binomial = new Abba.BinomialDistribution(numSamples, numSuccesses / numSamples);
}
Abba.Proportion.prototype = {
    // Compute an estimate of the underlying probability of success.
    pEstimate: function(zCriticalValue) {
        return this._adjustedWaldEstimate(zCriticalValue);
    },

    /* Generate an estimate for the underlying probability of success using the maximum likelihood
       estimator and the normal approximation error.  This is the so-called Wald interval:

       http://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Normal_approximation_interval
    */
    _waldEstimate: function() {
        return new Abba.ValueWithError(this._binomial.probability, this._binomial.standardError);
    },

    /* Compute the "adjusted Wald interval", which can be thought of as a Wald interval with
       (zCriticalValue^2 / 2) added to the number of successes and the number of failures. The
       estimated probability of success is the center of the interval. This provides much better
       coverage than the Wald interval (and many other intervals), though it has the unintuitive
       property that the estimated probabilty is not numSuccesses / numSamples. See (section 1.4.2
       and problem 1.24):

       Agresti, Alan. Categorical data analysis. New York, NY: John Wiley & Sons; 2002.
    */
    _adjustedWaldEstimate: function(zCriticalValue) {
        var squaredZCriticalValue = zCriticalValue * zCriticalValue;
        var adjustedNumSamples = this.numSamples + squaredZCriticalValue;
        var adjustedBinomial = new Abba.BinomialDistribution(
            adjustedNumSamples,
            (this.numSuccesses + squaredZCriticalValue / 2) / adjustedNumSamples);
        return new Abba.ValueWithError(
            adjustedBinomial.probability,
            adjustedBinomial.standardDeviation / adjustedBinomial.numSamples);
    },
};

Abba.ProportionComparison = function(baseline, trial, zCriticalValue) {
    this.baseline = baseline;
    this.trial = trial;
    this._zCriticalValue = zCriticalValue;
    this._standardNormal = new Abba.NormalDistribution();
}
Abba.ProportionComparison.prototype = {
    // Generate an estimate of the difference in success rates between the trial and the baseline.
    differenceEstimate: function() {
        var baselineP = this.baseline.pEstimate(this._zCriticalValue);
        var trialP = this.trial.pEstimate(this._zCriticalValue);
        var difference = trialP.value - baselineP.value;
        var standardError = Math.sqrt(Math.pow(baselineP.error, 2) + Math.pow(trialP.error, 2));
        return new Abba.ValueWithError(difference, standardError);
    },

    // Return the difference in sucess rates as a proportion of the baseline success rate.
    differenceRatio: function() {
        var baselineValue = this.baseline.pEstimate(this._zCriticalValue).value;
        var ratio = this.differenceEstimate().value / baselineValue;
        var error = this.differenceEstimate().error / baselineValue;
        return new Abba.ValueWithError(ratio, error);
    },

    /* Compute various values useful for comparing proportions with null hypothesis that they have
       the same probability of success
    */
    _computeTestValues: function() {
        var pooledProportion =
            (this.baseline.numSuccesses + this.trial.numSuccesses)
            / (this.baseline.numSamples + this.trial.numSamples);
        var expectedDifference =
            pooledProportion * (this.trial.numSamples - this.baseline.numSamples);
        var observedDifference = this.trial.numSuccesses - this.baseline.numSuccesses;
        return {
            pooledProportion: pooledProportion,
            expectedDifference: expectedDifference,
            varianceOfDifference:
                pooledProportion * (1 - pooledProportion)
                * (this.baseline.numSamples + this.trial.numSamples),
            observedAbsoluteDeviation: Math.abs(observedDifference - expectedDifference),
        };
    },

    /* For the given binomial distribution, compute an interval that covers at least
       (1 - coverageAlpha) of the total probability mass, centered at the expectation (unless we're
       at the boundary). Uses the normal approximation.
    */
    _binomialCoverageInterval: function(distribution, coverageAlpha) {
        if (distribution.numSamples < 1000) {
            // don't even bother trying to optimize for small-ish sample sizes
            return [0, distribution.numSamples];
        } else {
            return [
                Math.floor(distribution.inverseCdf(coverageAlpha / 2)),
                Math.ceil(distribution.inverseSurvival(coverageAlpha / 2)),
            ];
        }
    },

    /* Given the probability of an event, compute the probability that it happens at least once in
       numTimes independent trials. This is used to adjust a p-value for multiple comparisons.
       When used to adjust alpha instead, this is called a Sidak correction (the logic is the same,
       the formula is inverted):

       http://en.wikipedia.org/wiki/Bonferroni_correction#.C5.A0id.C3.A1k_correction
    */
    _probabilityUnion: function(probability, numTimes) {
        return 1 - Math.pow(1 - probability, numTimes);
    },

    /* Compute a p-value testing null hypothesis H0: pBaseline == pTrial against alternative
       hypothesis H1: pBaseline != pTrial by summing p-values conditioned on individual baseline
       success counts. This provides a more accurate correction for multiple testing but scales like
       O(sqrt(this.baseline.numSamples)), so can eventually get slow. In that case we fall back to
       zTest().

       Lower coverageAlpha increases accuracy at the cost of longer runtime. Roughly, the result
       will be accurate within no more than coverageAlpha (but this ignores error due to the normal
       approximation so isn't guaranteed).
    */
    iteratedTest: function(numTrials, coverageAlpha) {
        var values = this._computeTestValues();
        var trialDistribution = new Abba.BinomialDistribution(this.trial.numSamples,
                                                              values.pooledProportion);
        var baselineDistribution = new Abba.BinomialDistribution(this.baseline.numSamples,
                                                                 values.pooledProportion);

        // compute smallest and largest differences between success counts that are "at least as
        // extreme" as the observed difference (the observed difference is equal to one of these)
        var minExtremeDifference =
            Math.floor(values.expectedDifference - values.observedAbsoluteDeviation);
        var maxExtremeDifference =
            Math.ceil(values.expectedDifference + values.observedAbsoluteDeviation);

        var baselineLimits = this._binomialCoverageInterval(baselineDistribution, coverageAlpha);
        var pValue = 0;
        for (var baselineSuccesses = baselineLimits[0];
             baselineSuccesses <= baselineLimits[1];
             baselineSuccesses++) {
            // p-value of trial success counts "at least as extreme" for this particular baseline
            // success count
            var pValueAtBaseline =
                trialDistribution.cdf(baselineSuccesses + minExtremeDifference)
                + trialDistribution.survival(baselineSuccesses + maxExtremeDifference - 1);

            // this is exact because we're conditioning on the baseline count, so the multiple
            // trials are independent.
            var adjustedPValue = this._probabilityUnion(pValueAtBaseline, numTrials);

            var baselineProbability = baselineDistribution.mass(baselineSuccesses);
            pValue += baselineProbability * adjustedPValue;
        }

        // the remaining baseline values we didn't cover contribute less than coverageAlpha to the
        // sum, so adding that amount gives us a conservative upper bound.
        return pValue + coverageAlpha;
    },
};

// numTrials: number of trials to be compared to the baseline (i.e., not including the baseline)
Abba.Experiment = function(numTrials, baselineNumSuccesses, baselineNumSamples, baseAlpha) {
    normal = new Abba.NormalDistribution();
    this._baseline = new Abba.Proportion(baselineNumSuccesses, baselineNumSamples);

    this._numComparisons = Math.max(1, numTrials);
    // all z-values are two-tailed
    var baseZCriticalValue = normal.inverseSurvival(baseAlpha / 2);
    var alpha = baseAlpha / this._numComparisons // Bonferroni correction
    this._zCriticalValue = normal.inverseSurvival(alpha / 2);
    // z critical value for confidence interval on individual proportions. We compute intervals with
    // confidence level < alpha for individual trial proportions, so that they correspond neatly to
    // the confidence interval on the difference (which is computed at confidence level alpha). This
    // happens because some of the relative error disappears when we subtract the two proportions.
    this._trialIntervalZCriticalValue = this._zCriticalValue / Math.sqrt(2)
}
Abba.Experiment.prototype = {
    getBaselineProportion: function() {
        return this._baseline.pEstimate(this._trialIntervalZCriticalValue).valueWithInterval(
            this._trialIntervalZCriticalValue);
    },

    getResults: function(numSuccesses, numSamples) {
        var trial = new Abba.Proportion(numSuccesses, numSamples);
        var comparison = new Abba.ProportionComparison(
            this._baseline, trial, this._trialIntervalZCriticalValue);
        return {
            proportion: trial.pEstimate(this._trialIntervalZCriticalValue).valueWithInterval(
                this._trialIntervalZCriticalValue),
            relativeImprovement: comparison.differenceRatio().valueWithInterval(
                this._zCriticalValue),
            pValue: comparison.iteratedTest(this._numComparisons, 1e-5),
        };
    },
};

File: python/abba/stats.py
# Copyright (c) 2012 Thumbtack, Inc.

import collections
import math

from scipy import stats

def get_z_critical_value(alpha, two_tailed=True):
    """
    Returns the z critical value for a particular alpha = 1 - confidence level.  By default returns
    a two-tailed z-value, meaning the actual tail probability is alpha / 2.
    """
    if two_tailed:
        alpha /= 2
    return stats.distributions.norm.ppf(1 - alpha)

# a value with confidence interval bounds (not necessarily centered around the point estimate)
ValueWithInterval = collections.namedtuple(
    'ValueWithInterval',
    ('value', 'lower_bound', 'upper_bound'),
)

class ValueWithError(object):
    """
    A value with standard error, from which a confidence interval can be derived.
    """
    def __init__(self, value, error):
        self.value = value
        self.error = error

    def confidence_interval_width(self, z_critical_value):
        """
        z_critical_value should be the value at which the right-tail probability for a standard
        normal distribution equals half the desired alpha = 1 - confidence level:

        P(Z > z_value) = 1 - alpha / 2

        where Z is an N(0, 1) random variable.  Use get_z_critical_value(), or see
        http://en.wikipedia.org/wiki/Standard_normal_table.
        """
        return z_critical_value * self.error

    def value_with_interval(self, z_critical_value, estimated_value=None):
        width = self.confidence_interval_width(z_critical_value)
        return ValueWithInterval(
            value=estimated_value if estimated_value is not None else self.value,
            lower_bound=self.value - width,
            upper_bound=self.value + width,
        )

class BinomialDistribution(object):
    def __init__(self, num_trials, probability):
        self.num_trials = num_trials
        self.probability = probability
        self.expectation = num_trials * probability
        self.standard_deviation = math.sqrt(self.expectation * (1 - probability))
        self._binomial = stats.binom(num_trials, probability)

    def mass(self, count):
        return self._binomial.pmf(count)

    def cdf(self, count):
        return self._binomial.cdf(count)

    def survival(self, count):
        return 1 - self.cdf(count)

    def inverse_cdf(self, probability):
        return self._binomial.ppf(probability)

    def inverse_survival(self, probability):
        return self._binomial.isf(probability)

class Proportion(object):
    def __init__(self, num_successes, num_trials):
        """
        Represents a binomial proportion with num_successes successful samples out of num_trials
        total.
        """
        self.num_successes = num_successes
        self.num_trials = num_trials

    def p_estimate(self, z_critical_value=0):
        """
        Generate an adjusted estimate and error using the "Agresti-Coull Interval", see
        http://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Agresti-Coull_Interval.

        The estimated value is an adjusted best estimate for the actual probability. For example, if
        0 successes were observed out of 10 samples, it's unlikely the actual probability is zero,
        so the adjusted estimate will be slightly above zero.

        A z_critical_value of zero yields the ordinary Wald interval.
        """
        adjusted_num_trials = float(self.num_trials + z_critical_value**2)
        interval_center = (self.num_successes + z_critical_value**2 / 2) / adjusted_num_trials
        standard_error = math.sqrt(interval_center * (1 - interval_center) / adjusted_num_trials)
        return ValueWithError(interval_center, standard_error)

    def mixed_estimate(self, z_critical_value):
        """
        Returns an ValueWithInterval with a MLE value and upper/lower bounds from the Agresti-Coull
        interval.
        """
        return (
            self.p_estimate(z_critical_value=z_critical_value)
            .value_with_interval(z_critical_value, estimated_value=self.p_estimate().value)
        )

def confidence_interval_on_proportion(num_successes, num_trials, confidence_level=0.95):
    '''Convenience function with more straightforward interface.'''
    return Proportion(num_successes, num_trials).mixed_estimate(
        get_z_critical_value(1 - confidence_level)
    )

class ProportionComparison(object):
    def __init__(self, baseline, variation):
        self.baseline = baseline
        self.variation = variation

    def difference_estimate(self, z_critical_value):
        """
        Generate an estimate of the difference in success rates between the variation and the
        baseline.
        """
        baseline_p = self.baseline.p_estimate(z_critical_value=z_critical_value)
        variation_p = self.variation.p_estimate(z_critical_value=z_critical_value)
        difference = variation_p.value - baseline_p.value
        standard_error = math.sqrt(baseline_p.error ** 2 + variation_p.error ** 2)
        return ValueWithError(difference, standard_error)

    def difference_ratio(self, z_critical_value):
        """
        Return the difference in sucess rates as a proportion of the baseline success rate.
        """
        baseline_value = self.baseline.p_estimate(z_critical_value=z_critical_value).value
        difference = self.difference_estimate(z_critical_value=z_critical_value)
        ratio = difference.value / baseline_value
        error = difference.error / baseline_value
        return ValueWithError(ratio, error)

    def z_test(self, z_multiplier=1):
        """
        Perform a large-sample z-test of null hypothesis H0: p_baseline == p_variation against
        alternative hypothesis H1: p_baseline < p_variation.  Return the (one-tailed) p-value.

        z_multiplier: test z-value will be multiplied by this factor before computing a p-value.

        See http://en.wikipedia.org/wiki/Statistical_hypothesis_testing#Common_test_statistics,
        "Two-proportion z-test, pooled for d0 = 0".
        """
        pooled_stats = Proportion(
            self.baseline.num_successes + self.variation.num_successes,
            self.baseline.num_trials + self.variation.num_trials,
        )
        pooled_p_value = pooled_stats.p_estimate().value
        pooled_variance_of_difference = (
            pooled_p_value * (1 - pooled_p_value)
            * (1.0 / self.baseline.num_trials + 1.0 / self.variation.num_trials)
        )
        pooled_standard_error_of_difference = math.sqrt(pooled_variance_of_difference)
        test_z_value = self.difference_estimate(0).value / pooled_standard_error_of_difference
        adjusted_p_value = stats.distributions.norm.sf(test_z_value * z_multiplier)
        return adjusted_p_value

    def _binomial_coverage_interval(self, distribution, coverage_alpha):
        """
        For the given binomial distribution, compute an interval that covers at least (1 -
        coverage_alpha) of the total probability mass, centered at the expectation (unless we're at
        the boundary). Uses the normal approximation.
        """
        if distribution.num_trials < 1000:
            # don't even bother trying to optimize for small-ish sample sizes
            return (0, distribution.num_trials)
        else:
            return (
                int(math.floor(distribution.inverse_cdf(coverage_alpha / 2))),
                int(math.ceil(distribution.inverse_survival(coverage_alpha / 2))),
            )

    def _probability_union(self, probability, num_tests):
        """
        Given the probability of an event, compute the probability that it happens at least once in
        num_tests independent tests. This is used to adjust a p-value for multiple comparisons.
        When used to adjust alpha instead, this is called a Sidak correction (the logic is the same,
        the formula is inverted):
        http://en.wikipedia.org/wiki/Bonferroni_correction#.C5.A0id.C3.A1k_correction
        """
        return 1 - (1 - probability)**num_tests

    def iterated_test(self, num_tests, coverage_alpha, improvement_only=False):
        """
        Compute a p-value testing null hypothesis H0: p_baseline == p_variation against alternative
        hypothesis H1: p_baseline != p_variation by summing p-values conditioned on individual
        baseline success counts. This provides a more accurate correction for multiple testing but
        scales like O(sqrt(self.baseline.num_trials)), so can eventually get slow for very large
        values.

        Lower coverage_alpha increases accuracy at the cost of longer runtime. Roughly, the result
        will be accurate within no more than coverage_alpha (but this ignores error due to the
        normal approximation so isn't guaranteed).

        If improvement_only=True, computes p-value for alternative hypothesis
        H1: p_baseline < p_variation instead.
        """
        observed_delta = self.variation.p_estimate().value - self.baseline.p_estimate().value
        if observed_delta == 0 and not improvement_only:
            # a trivial case that the code below does not handle well
            return 1

        pooled_proportion = (
            (self.baseline.num_successes + self.variation.num_successes)
            / float(self.baseline.num_trials + self.variation.num_trials)
        )
        variation_distribution = BinomialDistribution(self.variation.num_trials, pooled_proportion)
        baseline_distribution = BinomialDistribution(self.baseline.num_trials, pooled_proportion)

        baseline_limits = self._binomial_coverage_interval(baseline_distribution, coverage_alpha)
        p_value = 0
        for baseline_successes in xrange(baseline_limits[0], baseline_limits[1] + 1):
            baseline_proportion = 1.0 * baseline_successes / self.baseline.num_trials
            if improvement_only:
                lower_trial_count = -1
                upper_trial_count = math.ceil(
                    (baseline_proportion + observed_delta) * self.variation.num_trials
                )
            else:
                observed_absolute_delta = abs(observed_delta)
                lower_trial_count = math.floor(
                    (baseline_proportion - observed_absolute_delta) * self.variation.num_trials
                )
                upper_trial_count = math.ceil(
                    (baseline_proportion + observed_absolute_delta) * self.variation.num_trials
                )

            # p-value of variation success counts "at least as extreme" for this particular
            # baseline success count
            p_value_at_baseline = (
                variation_distribution.cdf(lower_trial_count)
                + variation_distribution.survival(upper_trial_count - 1)
            )

            # this is exact because we're conditioning on the baseline count, so the multiple
            # tests are independent.
            adjusted_p_value = self._probability_union(p_value_at_baseline, num_tests)

            baseline_probability = baseline_distribution.mass(baseline_successes)
            p_value += baseline_probability * adjusted_p_value

        # the remaining baseline values we didn't cover contribute less than coverage_alpha to the
        # sum, so adding that amount gives us a conservative upper bound.
        return p_value + coverage_alpha

Results = collections.namedtuple(
    'Results',
    (
        'num_successes',
        'num_trials',
        'proportion', # ValueWithInterval
        'improvement', # ValueWithInterval
        'relative_improvement', # ValueWithInterval
        'two_tailed_p_value', # two-tailed p-value for trial != baseline
        'improvement_one_tailed_p_value', # one-tailed p-value for trial > baseline
    ),
)

class Experiment(object):
    P_VALUE_PRECISION = 1e-5

    def __init__(self, num_trials, baseline_num_successes, baseline_num_trials,
                 confidence_level=0.95):
        """
        num_trials: number of trials to be compared to the baseline
        confidence_level: used for all confidence intervals generated
        """
        self.num_comparisons = max(1, num_trials)
        self._baseline = Proportion(baseline_num_successes, baseline_num_trials)
        alpha = (1 - confidence_level) / num_trials # Bonferroni correction
        self._z_critical_value = get_z_critical_value(alpha)

    def get_baseline_proportion(self):
        return self._baseline.mixed_estimate(self._z_critical_value)

    def get_results(self, num_successes, num_trials):
        trial = Proportion(num_successes, num_trials)
        comparison = ProportionComparison(self._baseline, trial)
        return Results(
            num_successes=num_successes,
            num_trials=num_trials,
            proportion=trial.mixed_estimate(self._z_critical_value),
            improvement=comparison.difference_estimate(self._z_critical_value)
                .value_with_interval(
                    self._z_critical_value,
                    estimated_value=comparison.difference_estimate(0).value,
                ),
            relative_improvement=comparison.difference_ratio(self._z_critical_value)
                .value_with_interval(
                    self._z_critical_value,
                    estimated_value=comparison.difference_ratio(0).value,
                ),
            two_tailed_p_value=comparison.iterated_test(
                self.num_comparisons,
                self.P_VALUE_PRECISION,
            ),
            improvement_one_tailed_p_value=comparison.iterated_test(
                self.num_comparisons,
                self.P_VALUE_PRECISION,
                improvement_only=True,
            ),
        )
